Lib/importlib/_bootstrap_external.py

```diff
@@ -310,7 +320,18 @@ def cache_from_source(path, debug_override=None, *, optimization=None):
         if not optimization.isalnum():
             raise ValueError('{!r} is not alphanumeric'.format(optimization))
         almost_filename = '{}.{}{}'.format(almost_filename, _OPT, optimization)
-    return _path_join(head, _PYCACHE, almost_filename + BYTECODE_SUFFIXES[0])
+    filename = almost_filename + BYTECODE_SUFFIXES[0]
+    if bytecode_path is not None:
+        if not _path_isabs(head):
+            head = _path_join(_os.getcwd(), head)
```

Comment:
Is there a reason to suddenly start making this absolute? This is a change in semantics.

Reply:
It's not a change in semantics because this only occurs in the `if bytecode_path is not None` branch, which is new and only used if the new env var is set, so there are no existing semantics for this case.

We need to make the path absolute when using a `PYTHONBYTECODEPATH` to ensure no clashes in the cached bytecode tree. E.g if someone runs code with a local path of `foo/bar.py` in two different places on their system, with the same value for `PYTHONBYTECODEPATH`, we need them to store their bytecode in `/tmp/bytecode/abs/path1/to/foo/` and `/tmp/bytecode/abs/path2/to/foo/`, not collide over `/tmp/bytecode/foo/`.